### ISSUE TITLE
Questions About Using Proxygen

### ISSUE THREAD

**Issue by larrymeng** [facebook/proxygen#244]
**1) Thread Num**
eg:
HTTPServerOptions options;
options.threads = static_cast<size_t>(FLAGS_threads);
...

For the above example, does options.threads control the number of IO threads?

**2) Relationship between IO thread and RequestHandler**
Assuming that the number of IOthreads is 10, does it mean that the HTTP server can only handle up to 10 requests simultaneously?
Is there a RequestHandler in each IO thread that continuously handles requests from the client?

**3）Lifetime of RequestHandler**
eg:
class EchoHandlerFactory : public RequestHandlerFactory {
  ...

  RequestHandler* onRequest(RequestHandler*, HTTPMessage*) noexcept override {
    return new EchoHandler(stats_.get());
  }

 ...
};

For the above example, 
Under what circumstances, the onRequest function of Factory is called to generate a RequestHandler?
When will a RequestHandler be destroyed?
For each RequestHandler, is it processing one request(from the client) and then the next, or is it destroyed after processing one?

**4) Asynchronous processing in RequestHandler**
How to optimize the program if the RequestHandler takes longer processing time (such as MySQL or other IO operations)? 
Is there an asynchronous processing mechanism? eg:
onEOM() {
    auto eventBase = folly::EventBaseManager::get()->getEventBase();
    thread t([&, eventBase]() {

        auto d = fetch(); // call function to fetch data from db

        eventBase->runInEventBaseThread([&]() {
            sendResponse(d);
          });
});

 t.detach();
}

Any suggestions or code examples are welcome.

**Comment by udippant** [comment]
Hi @larrymeng ,

Per your questions:

> 1. Does options.threads control the number of IO threads? 


Yes. You can [see how that's used here](https://github.com/facebook/proxygen/blob/master/proxygen/httpserver/HTTPServer.cpp#L124)

> 2. Assuming that the number of IOthreads is 10, does it mean that the HTTP server can only handle up to 10 requests simultaneously? 


 Not necessarily. Proxygen is built with async 'event' driven model leveraging async libraries such as [folly eventbase library](https://github.com/facebook/folly/blob/master/folly/io/async/README.md). When designed carefully, the number of requests served should be order of magnitude higher. However, a lot of that depends you workload, nature of the requests, configuration, environment Proxygen is running on etc. You can get a good background info on this general issue here: http://www.kegel.com/c10k.html 

> 3. Under what circumstances, the onRequest function of Factory is called to generate a RequestHandler?


As mentioned in the proxygen's README, RequestHandlerFactory spawns a RequestHandler for each HTTP request.
The call chain is as follows:
 [`setupHeadersComplete` on Session calls the getRequestHandler(..)](https://github.com/facebook/proxygen/blob/master/proxygen/lib/http/session/HTTPDownstreamSession.cpp#L55) on the supplied impl of `SessionController`, which then calls ['newHandler(..)' on its RequestHandlerFactory impl (i.e. acceptor)](https://github.com/facebook/proxygen/blob/master/proxygen/lib/http/session/SimpleController.cpp#L24). It's from inside the implementation of `newHandler` [onRequest(..) is called](https://github.com/facebook/proxygen/blob/master/proxygen/httpserver/HTTPServerAcceptor.cpp#L111).

4. Typically, those operations are done on separate threads. Here's an issue that can be relevant to your question: https://github.com/facebook/proxygen/issues/240

**Comment by larrymeng** [comment]
Hi @udippant,
Thank you for your reply.
Your answer makes me very clear about the RequestHandler invocation chains.

those operations are done on separate threads.
eg:
void RequestHandler::onEOM()  noexcept {
  // give those separate worker threads the EventBase* from the RequestHandler thread
  folly::EventBase* eventBase = folly::EventBaseManager::get()->getEventBase();

  ... 

  // threadPool is folly::CPUThreadPoolExecutor/IOThreadPoolExecutor passed in from the handler factory
  threadPool->add([downstream_, eventBase]() {
    ... 
	// and then hand all the data back in a runInEventBaseThread from my worker.
    eventBase->runInEventBaseThread([&downstream_]() {
      ResponseBuilder(downstream_).status(...).header(...).body("").sendWithEOM();
    }   
  }
}

class XXXHandlerFactory : public RequestHandlerFactory {
  ...
    RequestHandler* onRequest(RequestHandler*, HTTPMessage*) noexcept override {
      return new XXXHandler(&threadPool);
    }

  private:
	// schedule the CPU-heavy work in separate worker threads
    folly::IOThreadPoolExecutor threadPool;
};


Is that correct?

**Comment by afrind** [comment]
The sample code looks correct.  Closing.
